Add Poetry `1.2.0` announcement (#68)
python-poetry/website: content/blog/2022-08-29-announcing-poetry-1-2-0.md

```bash
poetry install --only main
```

Finally, if you wish to install **only specific groups** of dependencies without installing the default dependencies,
the `--only` option can be used like so:

```bash
poetry install --only docs,test
```

#### Removing dependencies from a group

The [`remove`]({{< relref "docs/cli#remove" >}}) command supports a `--group` option to remove packages from a specific
group:

```bash
poetry remove mkdocs --group docs
```

### New plugin system

Poetry now supports a plugin system to alter or expand Poetry's functionality.

For example if your environment poses special requirements on the behaviour of Poetry which do not apply to the majority
of its users or if you wish to accomplish something with Poetry in a way that is not desired by most users.

In these cases you could consider creating a plugin to handle your specific logic.

#### Using plugins

Installed plugin packages are automatically loaded when Poetry starts up.

There are multiple ways to install plugins for Poetry, the easiest one (that should account for all the ways Poetry can
be installed) being:

```bash
poetry self add poetry-<PLUGIN_NAME>
```

> The `plugin add` command will ensure that the plugin is compatible with the current version of Poetry and install the

Reviewer: ```suggestion
The `self add` command will ensure that the plugin is compatible with the current version of Poetry and install the
```

Author: Woops, that's what happens when you only grep for `poetry plugin` :smile: 
Thanks!

needed packages for the plugin to work.

> The package specification formats supported by the `plugin add` command are the same as the ones supported by

Reviewer: ```suggestion
The package specification formats supported by the `self add` command are the same as the ones supported by
```

the [add command](/docs/cli/#add).

> If you no longer need a plugin and want to uninstall it, you can use the `plugin remove` command.

Reviewer: ```suggestion
If you no longer need a plugin and want to uninstall it, you can use the `self remove` command.
```

using the `--group (-G)` option.

```bash
$ poetry add pytest --group test
```

If the group does not already exist, it will be created automatically.

#### Installing group dependencies

**By default**, dependencies across **all groups** will be installed when executing `poetry install`.

You can **exclude** one or more groups with the `--without` option:

```bash
$ poetry install --without docs,test
```

You can also opt in [optional groups]({{< relref "#optional-groups" >}}) by using the `--with` option:

```bash
$ poetry install --with docs
```

If you only want to install the **default**, non-grouped, dependencies, you can do so with the `--only` option:

```bash
$ poetry install --only main
```

Finally, if you wish to install **only specific groups** of dependencies without installing the default dependencies,
the `--only` option can be used like so:

```bash
$ poetry install --only docs,test
```

#### Removing dependencies from a group

The [`remove`]({{< relref "docs/1.2/cli#remove" >}}) command supports a `--group` option to remove packages from a specific
group:

```bash
$ poetry remove mkdocs --group docs
```

> ### New plugin system

Reviewer: Not sure if we should put it into this section or someone else: We should mention that the `export` command is now provided via a plugin which is installed by default (but we plan to change this in a future version) and it is developed in its own repository with its own issue tracker..

Author: Yep, good idea! I won't be near a computer for most of the day, so in case you want to add a commit for that before I'm around, feel free to do so.

Reviewer: Done. Feel free to change it.

---
layout: single
title: "Announcing Poetry 1.2.0"
date: 2022-08-23 00:00:00
categories: [releases]
tags: ["1.x", "1.2"]
---

The Poetry team is pleased to announce the immediate availability of Poetry 1.2.0.

<!--more-->

Poetry 1.2 boasts a massive list of changes, new features and bugfixes developed over the course of 2 years, with

> contributions from dozens of contributors.

Reviewer 2: ```suggestion
changes from dozens of contributors.
```

Reviewer 3: now you repeat "changes" with "(...)a massive list of changes, new(...)"... I would leave it as it was before, with "contributions".

Reviewer 4: > now you repeat "changes" with "(...)a massive list of changes, new(...)"... I would leave it as it was before, with "contributions".

Or perhaps a slight tweak along the lines of: "… contributions from dozens of people." 🚲 🏡

If you have a previous version of Poetry installed via
the [new official installer]({{< relref "docs/1.2#installation" >}}), getting Poetry 1.2.0 is as easy as:

```bash
$ poetry self update
```

{{% warning %}}
If you installed Poetry using `get-poetry.py`, you will not be able to update to Poetry 1.2, since it can only be

> installed using a new installer detailed below.

Reviewer 2: ```suggestion
installed using the new installer detailed below.
```

> If you installed Poetry using `get-poetry.py`, you will not be able to update to Poetry 1.2, since it can only be
> installed using a new installer detailed below.
> 
> So you would first need to uninstall Poetry using:

Reviewer 2: ```suggestion
If you installed Poetry using the deprecated `get-poetry.py`, you will need to migrate to the new installer. Uninstall first with `get-poetry.py`:
```

Reviewer 3: "First, uninstall with `get-poetry.py`" would sound much better imho.

```bash
$ curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python3 - --uninstall
```

Then, follow the instructions below to install Poetry 1.2.
{{% /warning %}}

Since there are many changes in Poetry 1.2.0, this post details the changes over the following sections:

- [Breaking changes and major new features]({{< relref "#breaking-changes-and-major-new-features" >}})
- [Other noteworthy changes and new features]({{< relref "#other-noteworthy-changes-and-new-features" >}})
- [New commands]({{< relref "#new-commands" >}})
- [Changes in existing commands]({{< relref "#changes-in-existing-commands" >}})
- [New configuration options]({{< relref "#new-configuration-options" >}})
- [FAQ]({{< relref "#faq" >}})

For a complete list of changes, you can refer to the [change log](/history).

The full documentation for Poetry 1.2 is available [here]({{< relref "docs/1.2" >}}).

## Breaking changes and major new features

### New installation script

> `get-poetry.py` installation script has been replaced by a script hosted over https://install.python-poetry.org, whose
> source is available [here](https://github.com/python-poetry/install.python-poetry.org/blob/main/install-poetry.py).

Reviewer 2: ```suggestion
The legacy `get-poetry.py` installation script has been replaced by [install.python-poetry.org](https://install.python-poetry.org). The installer is now a [standalone project](https://github.com/python-poetry/install.python-poetry.org), with its own issue tracker.
```

{{% warning %}}

> `get-poetry.py` cannot install Poetry `>= 1.2.0`, so using https://install.python-poetry.org is required for newer
> versions.
> 

Reviewer 2: ```suggestion
```

Redundant

> `get-poetry.py` will still stay in Poetry's repository for some time, since many users rely on it, but since the new
> script can install Poetry `>= 1.1.7`, you can switch to the new installer even if still using 1.1.x.

Reviewer 2: ```suggestion
The `get-poetry.py` script is frozen, but will be available in Poetry's repository for at least one more minor release. However, the new installer can install Poetry `>= 1.1.7`, so all users should migrate away as soon as possible.
```

{{% /warning %}}

> Recommended way of installing Poetry (which is also detailed [here]({{< relref "docs/1.2#installation" >}})) is:

Reviewer 2: ```suggestion
Making use of the new installer is quite simple for most users ([full documentation]({{< relref "docs/1.2#installation" >}}) is also available):
```

their dependencies will be resolved and installed **by default**, they are simply a way to organize the dependencies
logically.
{{% /note %}}

The dependencies declared in `tool.poetry.dependencies` are part of an implicit `main` group.

```toml
[tool.poetry.dependencies]  # Implicit `main` dependency group
httpx = "*"
pendulum = "*"

[tool.poetry.group.test.dependencies]
pytest = "^7.1.0"
pytest-mock = "*"
```

Dependency groups, other than the implicit `main` group, must only contain dependencies you need in your development
process. Installing them is only possible by using Poetry.

To declare a set of dependencies, which add additional functionality to the project during runtime,
use [extras]({{< relref "docs/1.2/pyproject#extras" >}}) instead. Extras can be installed by the end user using `pip`.

{{% note %}}
**A note about the `dev-dependencies` section**

Any dependency declared in the `dev-dependencies` section will automatically be added to a `dev` group.
So the two following notations are equivalent:

```toml
[tool.poetry.dev-dependencies]
pytest = "^7.1.0"
pytest-mock = "*"
```

```toml
[tool.poetry.group.dev.dependencies]
pytest = "^7.1.0"
pytest-mock = "*"
```

> Poetry will slowly transition away from the `dev-dependencies` notation which will soon be deprecated, so it's advised

Reviewer: We can already call it "deprecated", can't we?

Author: Good call. Added https://github.com/python-poetry/website/pull/68/commits/2a6d0c0c777140e00919ac7f5efff9456a0be11e to update that.
